File: ros2/signal_bridge.cc
#include <algorithm>
#include <list>
#include <map>
#include <string>

#include "config/proto/config.pb.h"
#include "rclcpp/rclcpp.hpp"
#include "robot/action/proto/action_packet.pb.h"
#include "robot/perception/proto/perception_packet.pb.h"
#include "ros2/node_runner.h"
#include "ros2/proto/bridge.pb.h"
#include "ros2/proto/node.pb.h"
#include "ros2/proto/ros2_data_type.pb.h"
#include "ros2/utils/qos_setting.h"
#include "std_msgs/msg/u_int8_multi_array.hpp"

class SignalBridge : public rclcpp::Node {
 private:
  struct MappingRuntime {
    std::string mapping_id;
    std::string source_topic;
    std::string target_topic;
    ros2::bridge::PayloadType source_payload_type;
    ros2::bridge::MappingType mapping_type;
    ros2::bridge::AffineTransform affine;
    std::string action_id;
    rclcpp::Publisher<std_msgs::msg::UInt8MultiArray>::SharedPtr publisher;
    rclcpp::Subscription<std_msgs::msg::UInt8MultiArray>::SharedPtr subscription;
  };

 public:
  SignalBridge(const std::string& node_name, const int node_id, const config::Config& config)
      : Node(node_name) {
    for (const auto& single_bridge : config.robot().bridges().single_bridges()) {
      if (static_cast<int>(single_bridge.node().id()) != node_id ||
          single_bridge.node().node_type() != ros2::node::NodeType::SIGNAL_BRIDGE) {
        continue;
      }

      const auto qos = ros2_utils::CreateQosSetting(single_bridge.node().qos_setting());
      ValidateNodeTopics(single_bridge.node());

      for (const auto& mapping : single_bridge.bridge().mappings()) {
        if (mapping.source_topic().empty() || mapping.target_topic().empty()) {
          RCLCPP_ERROR(this->get_logger(),
                       "Mapping '%s' must define source_topic and target_topic.",
                       mapping.mapping_id().c_str());
          continue;
        }

        auto publisher_it = publishers_.find(mapping.target_topic());
        if (publisher_it == publishers_.end()) {
          auto publisher =
              this->create_publisher<std_msgs::msg::UInt8MultiArray>(mapping.target_topic(), qos);
          publisher_it = publishers_.emplace(mapping.target_topic(), publisher).first;
        }

        auto& runtime = mappings_.emplace_back(MappingRuntime{
            .mapping_id = mapping.mapping_id().empty()
                              ? mapping.source_topic() + "->" + mapping.target_topic()
                              : mapping.mapping_id(),
            .source_topic = mapping.source_topic(),
            .target_topic = mapping.target_topic(),
            .source_payload_type = mapping.source_payload_type(),
            .mapping_type = mapping.mapping_type(),
            .affine = mapping.affine(),
            .action_id = mapping.action_id(),
            .publisher = publisher_it->second,
        });

        runtime.subscription = this->create_subscription<std_msgs::msg::UInt8MultiArray>(
            runtime.source_topic,
            qos,
            [this, &runtime](const std_msgs::msg::UInt8MultiArray::ConstSharedPtr msg) {
              ProcessMessage(runtime, *msg);
            });
      }
    }

    if (mappings_.empty()) {
      RCLCPP_ERROR(this->get_logger(), "No signal bridge mappings found for node_id %d!", node_id);
      return;
    }

    RCLCPP_INFO(this->get_logger(),
                "Signal bridge started with %zu mapping(s) for node_id %d.",
                mappings_.size(),
                node_id);
  }

 private:
  std::list<MappingRuntime> mappings_;
  std::map<std::string, rclcpp::Publisher<std_msgs::msg::UInt8MultiArray>::SharedPtr> publishers_;

  void ValidateNodeTopics(const ros2::node::Node& node_cfg) {
    for (const auto& sub : node_cfg.subscriptions()) {
      if (sub.ros2_data_type() != ros2::data_type::UINT8_MULTI_ARRAY) {
        RCLCPP_ERROR(this->get_logger(),
                     "Unsupported subscription type %d on topic '%s'. "
                     "Signal bridge requires UINT8_MULTI_ARRAY.",
                     static_cast<int>(sub.ros2_data_type()),
                     sub.topic().c_str());
      }
    }
    for (const auto& pub : node_cfg.publishers()) {
      if (pub.ros2_data_type() != ros2::data_type::UINT8_MULTI_ARRAY) {
        RCLCPP_ERROR(this->get_logger(),
                     "Unsupported publisher type %d on topic '%s'. "
                     "Signal bridge requires UINT8_MULTI_ARRAY.",
                     static_cast<int>(pub.ros2_data_type()),
                     pub.topic().c_str());
      }
    }
  }

  void ProcessMessage(const MappingRuntime& runtime, const std_msgs::msg::UInt8MultiArray& msg) {
    robot::action::ActionPacket action;
    if (!ToActionPacket(runtime, msg, action)) {
      return;
    }
    auto serialized = action.SerializeAsString();
    std_msgs::msg::UInt8MultiArray out;
    out.data.assign(serialized.begin(), serialized.end());
    runtime.publisher->publish(out);
  }

  bool ToActionPacket(const MappingRuntime& runtime,
                      const std_msgs::msg::UInt8MultiArray& msg,
                      robot::action::ActionPacket& out) {
    if (runtime.source_payload_type == ros2::bridge::PAYLOAD_TYPE_ACTION_PACKET) {
      if (!out.ParseFromArray(msg.data.data(), static_cast<int>(msg.data.size()))) {
        RCLCPP_ERROR(this->get_logger(),
                     "Failed to parse ActionPacket for mapping '%s'.",
                     runtime.mapping_id.c_str());
        return false;
      }
      if (runtime.mapping_type != ros2::bridge::MAPPING_TYPE_PASS_THROUGH_ACTION &&
          runtime.mapping_type != ros2::bridge::MAPPING_TYPE_INVALID) {
        RCLCPP_WARN(this->get_logger(),
                    "Mapping '%s' uses ACTION_PACKET source with mapping_type %d. "
                    "Pass-through is applied.",
                    runtime.mapping_id.c_str(),
                    static_cast<int>(runtime.mapping_type));
      }
      if (!runtime.action_id.empty()) {
        out.set_action_id(runtime.action_id);
      }
      return true;
    }

    if (runtime.source_payload_type == ros2::bridge::PAYLOAD_TYPE_PERCEPTION_PACKET) {
      robot::perception::PerceptionPacket perception;
      if (!perception.ParseFromArray(msg.data.data(), static_cast<int>(msg.data.size()))) {
        RCLCPP_ERROR(this->get_logger(),
                     "Failed to parse PerceptionPacket for mapping '%s'.",
                     runtime.mapping_id.c_str());
        return false;
      }
      if (runtime.mapping_type != ros2::bridge::MAPPING_TYPE_POSITION_TO_POSITION) {
        RCLCPP_ERROR(this->get_logger(),
                     "Unsupported mapping_type %d for PERCEPTION_PACKET in mapping '%s'.",
                     static_cast<int>(runtime.mapping_type),
                     runtime.mapping_id.c_str());
        return false;
      }
      if (!perception.has_position()) {
        RCLCPP_WARN(this->get_logger(),
                    "PerceptionPacket in mapping '%s' has no position field.",
                    runtime.mapping_id.c_str());
        return false;
      }

      float value = perception.position().position();
      if (runtime.affine.invert()) {
        value = -value;
      }
      float scale = 1.0f;
      float offset = 0.0f;
      if (runtime.affine.ByteSizeLong() > 0) {
        scale = runtime.affine.scale();
        offset = runtime.affine.offset();
      }
      value = value * scale + offset;
      if (runtime.affine.clamp_enabled()) {
        value = std::max(runtime.affine.clamp_min(), std::min(runtime.affine.clamp_max(), value));
      }

      out.Clear();
      out.set_position(value);
      out.set_timestamp_ns(perception.timestamp_ns() != 0
                               ? perception.timestamp_ns()
                               : static_cast<int64_t>(this->get_clock()->now().nanoseconds()));
      out.set_action_id(runtime.action_id.empty() ? runtime.mapping_id : runtime.action_id);
      return true;
    }

    RCLCPP_ERROR(this->get_logger(),
                 "Unsupported source_payload_type %d in mapping '%s'.",
                 static_cast<int>(runtime.source_payload_type),
                 runtime.mapping_id.c_str());
    return false;
  }
};

int main(int argc, char* argv[]) {
  return ros2_utils::RunNode<SignalBridge>(argc, argv, "signal_bridge");
}

File: ros2/signal_bridge.py
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Dict

from rclpy.node import Node
from std_msgs.msg import UInt8MultiArray

from config.proto import config_pb2
from robot.action.proto import action_packet_pb2
from robot.perception.proto import perception_packet_pb2
from ros2.node_runner import run_node
from ros2.proto import bridge_pb2, node_pb2, ros2_data_type_pb2
from ros2.utils.qos_setting import create_qos_setting


@dataclass
class MappingRuntime:
    mapping_id: str
    source_topic: str
    target_topic: str
    source_payload_type: int
    mapping_type: int
    action_id: str
    affine: bridge_pb2.AffineTransform
    publisher: object
    subscription: object


class SignalBridge(Node):
    def __init__(self, node_name: str, node_id: int, config: config_pb2.Config):
        super().__init__(node_name)
        self._mappings: Dict[str, MappingRuntime] = {}
        self._publishers: Dict[str, object] = {}

        for single_bridge in config.robot.bridges.single_bridges:
            if (
                int(single_bridge.node.id) != node_id
                or single_bridge.node.node_type != node_pb2.NodeType.SIGNAL_BRIDGE
            ):
                continue

            qos = create_qos_setting(single_bridge.node.qos_setting)
            self._validate_node_topics(single_bridge.node)

            for mapping in single_bridge.bridge.mappings:
                if not mapping.source_topic or not mapping.target_topic:
                    self.get_logger().error(
                        "Mapping '%s' must define source_topic and target_topic.",
                        mapping.mapping_id,
                    )
                    continue

                publisher = self._publishers.get(mapping.target_topic)
                if publisher is None:
                    publisher = self.create_publisher(
                        UInt8MultiArray,
                        mapping.target_topic,
                        qos,
                    )
                    self._publishers[mapping.target_topic] = publisher

                runtime = MappingRuntime(
                    mapping_id=mapping.mapping_id
                    or f"{mapping.source_topic}->{mapping.target_topic}",
                    source_topic=mapping.source_topic,
                    target_topic=mapping.target_topic,
                    source_payload_type=mapping.source_payload_type,
                    mapping_type=mapping.mapping_type,
                    action_id=mapping.action_id,
                    affine=mapping.affine,
                    publisher=publisher,
                    subscription=None,
                )

                runtime.subscription = self.create_subscription(
                    UInt8MultiArray,
                    runtime.source_topic,
                    self._make_callback(runtime),
                    qos,
                )
                self._mappings[runtime.mapping_id] = runtime

        if not self._mappings:
            self.get_logger().error(
                "No signal bridge mappings found for node_id %d!", node_id
            )
            return

        self.get_logger().info(
            "Signal bridge started with %d mapping(s) for node_id %d.",
            len(self._mappings),
            node_id,
        )

    def _validate_node_topics(self, node_cfg: node_pb2.Node) -> None:
        for sub in node_cfg.subscriptions:
            if sub.ros2_data_type != ros2_data_type_pb2.UINT8_MULTI_ARRAY:
                self.get_logger().error(
                    "Unsupported subscription type %s on topic '%s'. "
                    "Signal bridge requires UINT8_MULTI_ARRAY.",
                    str(sub.ros2_data_type),
                    sub.topic,
                )
        for pub in node_cfg.publishers:
            if pub.ros2_data_type != ros2_data_type_pb2.UINT8_MULTI_ARRAY:
                self.get_logger().error(
                    "Unsupported publisher type %s on topic '%s'. "
                    "Signal bridge requires UINT8_MULTI_ARRAY.",
                    str(pub.ros2_data_type),
                    pub.topic,
                )

    def _make_callback(self, runtime: MappingRuntime):
        def callback(msg: UInt8MultiArray):
            action = self._to_action_packet(runtime, msg)
            if action is None:
                return
            out = UInt8MultiArray()
            out.data = action.SerializeToString()
            runtime.publisher.publish(out)

        return callback

    def _to_action_packet(
        self, runtime: MappingRuntime, msg: UInt8MultiArray
    ) -> action_packet_pb2.ActionPacket | None:
        if runtime.source_payload_type == bridge_pb2.PAYLOAD_TYPE_ACTION_PACKET:
            action = action_packet_pb2.ActionPacket()
            try:
                action.ParseFromString(bytes(msg.data))
            except Exception as exc:
                self.get_logger().error(
                    "Failed to parse ActionPacket for mapping '%s': %s",
                    runtime.mapping_id,
                    str(exc),
                )
                return None
            if runtime.mapping_type not in (
                bridge_pb2.MAPPING_TYPE_PASS_THROUGH_ACTION,
                bridge_pb2.MAPPING_TYPE_INVALID,
            ):
                self.get_logger().warning(
                    "Mapping '%s' uses ACTION_PACKET source with mapping_type %s. "
                    "Pass-through is applied.",
                    runtime.mapping_id,
                    str(runtime.mapping_type),
                )
            if runtime.action_id:
                action.action_id = runtime.action_id
            return action

        if runtime.source_payload_type == bridge_pb2.PAYLOAD_TYPE_PERCEPTION_PACKET:
            perception = perception_packet_pb2.PerceptionPacket()
            try:
                perception.ParseFromString(bytes(msg.data))
            except Exception as exc:
                self.get_logger().error(
                    "Failed to parse PerceptionPacket for mapping '%s': %s",
                    runtime.mapping_id,
                    str(exc),
                )
                return None

            if runtime.mapping_type != bridge_pb2.MAPPING_TYPE_POSITION_TO_POSITION:
                self.get_logger().error(
                    "Unsupported mapping_type %s for PERCEPTION_PACKET in mapping '%s'.",
                    str(runtime.mapping_type),
                    runtime.mapping_id,
                )
                return None

            if not perception.HasField("position"):
                self.get_logger().warning(
                    "PerceptionPacket in mapping '%s' has no position field.",
                    runtime.mapping_id,
                )
                return None

            value = float(perception.position.position)
            if runtime.affine.invert:
                value = -value
            # Use identity transform when affine block is omitted in config.
            scale = (
                1.0
                if runtime.affine == bridge_pb2.AffineTransform()
                else runtime.affine.scale
            )
            value = value * scale + runtime.affine.offset
            if runtime.affine.clamp_enabled:
                value = max(
                    runtime.affine.clamp_min, min(runtime.affine.clamp_max, value)
                )

            action = action_packet_pb2.ActionPacket()
            action.position = value
            action.timestamp_ns = (
                perception.timestamp_ns
                if perception.timestamp_ns
                else int(time.time_ns())
            )
            action.action_id = runtime.action_id or runtime.mapping_id
            return action

        self.get_logger().error(
            "Unsupported source_payload_type %s in mapping '%s'.",
            str(runtime.source_payload_type),
            runtime.mapping_id,
        )
        return None


def main() -> int:
    return run_node(SignalBridge, "signal_bridge")


if __name__ == "__main__":
    raise SystemExit(main())
